### PR TITLE
fix(kafka): Use librdkafka defaults for queue sizes unless overridden

### PR DESCRIPTION
**Pull request by tkaemming**
`queued.max.messages.kbytes` and `queued.min.messages` will now be resolved in order of precedence (if a value is not `None` it will be used, otherwise the next location is checked):

1. command line argument
2. settings value
3. librdkafka default (this is new)

### CODE REVIEW

File: snuba/settings_base.py [hunk 1]
```diff
@@ -56,8 +56,8 @@
 
 DEFAULT_MAX_BATCH_SIZE = 50000
 DEFAULT_MAX_BATCH_TIME_MS = 2 * 1000
-DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 10000
-DEFAULT_QUEUED_MIN_MESSAGES = 10000
+DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = None  # use librdkafka default
+DEFAULT_QUEUED_MIN_MESSAGES = None  # use librdkafka default
```
Comment (fpacifici) [inline snuba/settings_base.py]: why not removing them altogether ? Is it to still be able to make a global change ?

**Review by fpacifici — APPROVED** [review]

File: snuba/settings_base.py (hunk 1 above)
Comment (tkaemming) [reply snuba/settings_base.py]: Yeah, I think it is still beneficial to be able to change the default value for this configuration without having to specify it in each command line invocation.

**Comment by tkaemming** [comment]
Added @mattrobenolt and @JTCunning since this potentially affects single tenant and production configurations.

**Comment by mattrobenolt** [comment]
@b1naryth1ef

**Comment by tkaemming** [comment]
I guess @BYK should know about this too.

File: snuba/settings_base.py [hunk 2]
```diff
@@ -56,8 +56,8 @@
 
 DEFAULT_MAX_BATCH_SIZE = 50000
 DEFAULT_MAX_BATCH_TIME_MS = 2 * 1000
-DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 10000
-DEFAULT_QUEUED_MIN_MESSAGES = 10000
```
Comment (tkaemming) [inline snuba/settings_base.py]: The librdkafka defaults for these are 1048576 and 100000, respectively. They are documented [here](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md) as `queued.max.messages.kbytes` and `queued.min.messages`.

**Comment by tkaemming** [comment]
Also, for everyone else's context (namely @BYK and @b1naryth1ef) — we ran into an issue today where consumers were backlogging, but not saturating CPU or IO. It turns out, when the (per partition) queue size limit is set to 10MB, and the maximum size of a Sentry event is 10MB, the round trip time between the consumer and the broker becomes the limiting factor. The 10MB limit is less than 1% of the default Kafka configuration for this value. None of us really expected this value to be so different than the default value. (It isn't quite so different any longer.)

I think it's going to be difficult to give a reasonable default to this value with limited information (or in the case of on-premise: potentially zero information) about the hardware the consumer is actually running on, and in that case, it's best just to use the (potentially unreasonable) librdkafka default of ~1GB for the project in general. (I'd also like to eventually allow broader control over the configuration parameters passed to librdkafka instead of limiting the tuning to the few configuration parameters we've explicitly exposed, but that will be a larger change that happens in the far future.)

The main concern with this setting is that if you have many partitions (again, this depends on your specific configuration: by default, [most users will have one](https://github.com/getsentry/snuba/blob/28830c58640b8f98505c11ded2c381798a88708d/snuba/datasets/table_storage.py#L19) unless they manually create the topic) and a large backlog, the consumer will consume too much memory and be OOM killed (which probably causes the backlog to grow further, putting the consumers into a vicious cycle.)

Since this queue size is per partition, you probably want to figure out the appropriate value by doing something like: `available memory for a single consumer / (total number of partitions / total number of consumers)`, leaving room for interpreter overhead, etc. as well as a safety margin for a consumer to accept extra partitions during rebalances.

For single tenant, we can probably be more aggressive with this to get better performance based on the available resources.

For on-premise, we should probably set this value very conservatively to avoid lots of "my consumers keep crashing and where are my events" GitHub issues from people trying to run on low-budget hardware, and deal with the fact that throughput will be throttled unless the user takes it upon themselves to change the configuration to reflect their available resources.

**Comment by BYK** [comment]
@tkaemming on on-premise, we simply use the default values for these. Should we be changing them (sounds like it based on your last message)? If so, I need guidance. Your comments gives me an idea but I still am not familiar with running Kafka enough to pick a value based on that context.

**Comment by BYK** [comment]
Feel like this got quite stale and should be closed?

**Comment by fpacifici** [comment]
These parameters are now configurable per consumer. 
https://github.com/getsentry/snuba/blob/master/snuba/cli/consumer.py#L74-L85

Other parameters are also configurable per broker:
https://github.com/getsentry/snuba/blob/master/snuba/settings/__init__.py#L116-L126

So hardcoded parameters should not be a problem anymore.

**Pull request closed without merging**